src/Std/Data/DTreeMap/Internal/Lemmas.lean

```diff
@@ -1465,4 +1466,109 @@ theorem getThenInsertIfNew?!_snd [TransOrd α] (h : t.WF) {k : α} {v : β} :
 
 end Const
 
+theorem length_keys [TransOrd α] (h : t.WF) :
+    t.keys.length = t.size := by
+  simp_to_model using List.length_keys_eq_length
+
+theorem isEmpty_keys :
+    t.keys.isEmpty = t.isEmpty  := by
+  simp_to_model using List.isEmpty_keys_eq_isEmpty
+
+theorem contains_keys [instBEq : BEq α] [LawfulBEq α] [LawfulBEqOrd α] [TransOrd α] {k : α} (h : t.WF) :
+    t.keys.contains k = t.contains k := by
+  rw [contains_eq_containsKey h.ordered, ← eq_beqOfOrd_of_lawfulBEqOrd]
+  simp_to_model using (List.containsKey_eq_keys_contains (a := k) (l := t.toListModel)).symm
+
+theorem mem_keys [LawfulEqOrd α] [TransOrd α] {k : α} (h : t.WF) :
+    k ∈ t.keys ↔ k ∈ t := by
+  simpa only [mem_iff_contains, ← List.contains_iff, ← Bool.eq_iff_iff] using contains_keys h
+
+theorem distinct_keys [TransOrd α] (h : t.WF) :
+    t.keys.Pairwise (fun a b => ¬ compare a b = .eq) := by
+  simp only [← not_congr beq_iff_eq, ← beq_eq, Bool.not_eq_true]
+  simp_to_model using h.ordered.distinctKeys.distinct
+
+theorem map_sigma_fst_toList_eq_keys :
+    t.toList.map Sigma.fst = t.keys := by
+  simp_to_model using (List.keys_eq_map ..).symm
+
+theorem length_toList [TransOrd α] (h : t.WF) :
+    t.toList.length = t.size := by
+  simp_to_model
+
+theorem isEmpty_toList :
+    t.toList.isEmpty = t.isEmpty := by
+  simp_to_model
+
+theorem mem_toList_iff_get?_eq_some [TransOrd α] [LawfulEqOrd α] {k : α} {v : β k} (h : t.WF) :
+    ⟨k, v⟩ ∈ t.toList ↔ t.get? k = some v := by
+  simp_to_model using List.mem_iff_getValueCast?_eq_some
+
+theorem find?_toList_eq_some_iff_get?_eq_some [TransOrd α] [LawfulEqOrd α] {k : α} {v : β k}
+    (h : t.WF) :
+    t.toList.find? (compare ·.1 k == .eq) = some ⟨k, v⟩ ↔ t.get? k = some v := by
+  simp_to_model using List.find?_eq_some_iff_getValueCast?_eq_some
+
+theorem find?_toList_eq_none_iff_contains_eq_false [TransOrd α] {k : α} (h : t.WF) :
+    t.toList.find? (compare ·.1 k == .eq) = none ↔ t.contains k = false := by
+  simp_to_model using List.find?_eq_none_iff_containsKey_eq_false
+
+theorem find?_toList_eq_none_iff_not_mem [TransOrd α] {k : α} (h : t.WF) :
+    t.toList.find? (·.1 == k) = none ↔ ¬ k ∈ t := by
```

Comment:
Is this using the `BEq` instance derived from the `Ord` instance? Should this be
```suggestion
    t.toList.find? (compare ·.1 k == .eq) = none ↔ ¬ k ∈ t := by
```
like above?

Reply:
Yes and yes, it's definitionally equivalent but using `compare` is the better formulation. :+1: